sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala

```diff
@@ -0,0 +1,208 @@
+    service.port = actualPort
+    service
+  }
+
+}
+
+/**
+ * Implements the EventingTestKit protocol defined in protocols/testkit/src/main/protobuf/eventing_test_backend.proto
+ */
+final class EventingTestServiceImpl(system: ActorSystem, val host: String, var port: Int, codec: MessageCodec)
+    extends EventingTestKit {
+
+  private val log = LoggerFactory.getLogger(classOf[EventingTestKit])
+  private implicit val sys = system
+
+  private val eventDestinationProbe = new ConcurrentHashMap[EventDestination, TestProbe]()
+
+  override def getTopic(topic: String): Topic = {
+    val dest = EventDestination.defaultInstance.withTopic(topic)
+    val probe = eventDestinationProbe.computeIfAbsent(dest, _ => TestProbe())
+    new TopicImpl(probe, codec)
+  }
+
+  final class ServiceImpl extends EventingTestKitService {
+    override def emitSingle(in: EmitSingleCommand): Future[EmitSingleResult] = {
+      log.info("emitSingle request: [{}]", in)
+
+      in.destination.foreach(dest => {
+        val probe = eventDestinationProbe.computeIfAbsent(dest, _ => TestProbe())
+        probe.ref ! in
+      })
+
+      if (in.destination.isEmpty) {
+        log.warn("Received a message without destination, ignoring. {}", in)
+      }
+      Future.successful(EmitSingleResult())
+    }
+
+    override def runSource(in: Source[RunSourceCommand, NotUsed]): Source[SourceElem, NotUsed] = ???
+
+    override def eventStreamOut(in: Source[EventStreamOutCommand, NotUsed]): Source[EventStreamOutResult, NotUsed] =
+      throw new UnsupportedOperationException("Feature not supported in the testkit yet")
+  }
+}
+
+class TopicImpl(private val probe: TestProbe, codec: MessageCodec) extends Topic {
+
+  override def expectNone(): Unit = expectNone(DefaultTimeout)
+
+  override def expectNone(timeout: time.Duration): Unit = probe.expectNoMessage(timeout.toScala)
+
+  override def expectOneRaw(): TestKitMessage[ByteString] = expectOneRaw(DefaultTimeout)
+
+  override def expectOneRaw(timeout: time.Duration): TestKitMessage[ByteString] = {
+    val msg = probe.expectMsgType[EmitSingleCommand](timeout.toScala)
+    TestKitMessageImpl.ofMessage(msg.getMessage)
+  }
+
+  override def expectOne(): TestKitMessage[_] = expectOne(DefaultTimeout)
+
+  override def expectOne(timeout: time.Duration): TestKitMessage[_] = {
+    val msg = probe.expectMsgType[EmitSingleCommand](timeout.toScala)
+    anyFromMessage(msg.getMessage)
+  }
+
+  override def expectMessageType[T <: GeneratedMessageV3](clazz: Class[T]): TestKitMessage[T] =
+    expectMessageType(clazz, DefaultTimeout)
+
+  override def expectMessageType[T <: GeneratedMessageV3](
+      clazz: Class[T],
+      timeout: time.Duration): TestKitMessage[T] = {
+    val msg = probe.expectMsgType[EmitSingleCommand]
+    val metadata = new MetadataImpl(msg.getMessage.getMetadata.entries)
+    val decodedMsg = codec.decodeMessage(ScalaPbAny(metadata.get("ce-type").orElse(""), msg.getMessage.payload))
```

Comment:
What would happen with an empty type? Would it fail or sth?

Reply:
yeah, it fails.. I think I will add a TopicImplSpec to document/test such and other scenarios.